src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -602,20 +602,29 @@ void C2_MacroAssembler::fast_lock(Register objReg, Register boxReg, Register tmp
   jccb(Assembler::notZero, IsInflated);
 
   if (!UseHeavyMonitors) {
-    // Attempt stack-locking ...
-    orptr (tmpReg, markWord::unlocked_value);
-    movptr(Address(boxReg, 0), tmpReg);          // Anticipate successful CAS
-    lock();
-    cmpxchgptr(boxReg, Address(objReg, oopDesc::mark_offset_in_bytes()));      // Updates tmpReg
-    jcc(Assembler::equal, COUNT);           // Success
-
-    // Recursive locking.
-    // The object is stack-locked: markword contains stack pointer to BasicLock.
-    // Locked by current thread if difference with current SP is less than one page.
-    subptr(tmpReg, rsp);
-    // Next instruction set ZFlag == 1 (Success) if difference is less then one page.
-    andptr(tmpReg, (int32_t) (NOT_LP64(0xFFFFF003) LP64_ONLY(7 - (int)os::vm_page_size())) );
-    movptr(Address(boxReg, 0), tmpReg);
+    if (UseFastLocking) {
+      Label slow;
+      fast_lock_impl(objReg, tmpReg, thread, scrReg, slow);
+      jmp(COUNT);
+      bind(slow);
+      testptr(objReg, objReg); // force ZF=0 to indicate failure
+      jmp(NO_COUNT);
```

Comment:
We set a flag on failure (`NO_COUNT`) path. Should we set the flag on success (`COUNT`) path as well?

Reply:
The path at COUNT already sets the ZF, there is no need to do it here. NO_COUNT doesn't clear ZF, and fast_lock_impl() may branch to slow with ZF set (on the overflow check) so we need to explicitly clear ZF. However, I just came up with a better way to check for overflow that readily clears the ZF: subtract 1 from the end-offset and make a greater-comparison instead of the greaterEquals that we currently do on the end-offset. That should simplify the code and avoid a branch.